### DISCUSSION TITLE
Logica on Local Jupyter

### DISCUSSION THREAD

**Discussion by cdyellick** [EvgSkv/logica#212] [Q&A] [answered]
I’m interested in using Logica with locally installed Postgres and Jupyter.  All the examples start by importing colab_logica from logica.  Not surprisingly, it depends on some google packages that I do not have.

Is it possible to run logica in a locally hosted Jupyter notebook?  If so, are there any examples available?

**Comment by EvgSkv** [comment, marked as answer]
Hi @cdyellick 
google packages are supposed to be behind try/except and colab_logica _should_ work fine without them, letting you use Postgres.
Here is an example of running Logica with SQLite on Kaggle (which doesn't have Google libraries):
https://www.kaggle.com/evgenyskvortsov/logica
And here is an example of running PosgreSQL:
https://www.kaggle.com/evgenyskvortsov/logica-postgres

Let me know if you have any further questions, or encounter any problems running Logica youself.

**Reply by cdyellick** [reply]
Thank you for the quick reply @EvgSkv.  I'm not sure what I was doing wrong before, but that got me over the hump.  Sometimes it helps just to know it;s possible.

BTW.  This is a very interesting project and I appreciate you being so approachable.  I'm new to logic programming, but the potential of something that is more readable than SQL has me very intrigued.  Enough so, that I picked up a Prolog book to build a foundational knowledge of logic programming.  Looking forward to spending some time with Logica.

**Reply by EvgSkv** [reply]
Happy to help! Don't hesitate to reach out if you have further questions.

**Comment by wdduncan** [comment]
@cdyellick I went to the kaggle sites you mention, but the scripts were not there. Did you take them down?

**Reply by cdyellick** [reply]
@wdduncan They are actually not my notebooks, I believe they belong to @EvgSkv.  If I recall, it took me an extra click or two to get them to load when I followed the links, but trying them again now, its not working for me either.

**Reply by wdduncan** [reply]
Thanks @cdyellick ... pinging @EvgSkv  :)   
BTW, I hope you keep the project going. I think this is really cool!

**Reply by EvgSkv** [reply]
Sorry I missed the question!

It looks like I need to "save a version" on Kaggle for it to be visible.
Please try this link:
https://www.kaggle.com/code/evgenyskvortsov/logica?scriptVersionId=97329668

Please let me know if it opens this time! :-)

**Reply by wdduncan** [reply]
Yes. I can open the link now :)  
Although, it does show the messages:
```
Could not import google.cloud.auth.
Could not import google.colab.widgets.
```
I don't know if that is important.

**Reply by EvgSkv** [reply]
Those messages are expected: they are basically informing you that Google Cloud libraries weren't available, so BigQuery can't be used. In this mode you can use SQLite. If you are running your own back-end with Google Cloud libraries then BQ should work too.

**Reply by yilinxia** [reply]
Hi @EvgSkv the SQLite engine is now operational on my local pc. Based on your earlier response, I should implement my own backend using Google Cloud Libraries. Could you elaborate on how to set that up? tks

**Reply by EvgSkv** [reply]
Hi @yilinxia 
there is nothing to implement really. You'd just need to install Google Cloud SDK on your machine and register Google Cloud Project ID.
Furthermore, you can simply use CoLab and then no installation is needed: https://colab.corp.google.com/
Here is a doc about creating Google Cloud Projects: https://cloud.google.com/resource-manager/docs/creating-managing-projects
Let me know if you have further questions!

**Reply by yilinxia** [reply]
Hi, @EvgSkv Thanks for your explanation. I just set up the Google Cloud however, I encountered some unexpected errors shown below. Do you know why this happened?

no error raised for this part
```
import os
os.environ["GOOGLE_APPLICATION_CREDENTIALS"] = 'path_to_the_json_file'
from google.colab import auth
from logica import colab_logica
from IPython.utils import traitlets as _traitlets
```
input
```
%%logica Employee

Employee(name: "Alice", role: "Product Manager");
Employee(name: "Bob", role: "Engineer");
Employee(name: "Caroline", role: "Engineer");
Employee(name: "David", role: "Data Scientist");
Employee(name: "Eve", role: "Data Scientist");
```

output
```
Javascript Error: colab_lib is not defined
Javascript Error: window.id1.setSelectedTabIndex is not a function
Javascript Error: google is not defined
Javascript Error: google is not defined
Javascript Error: window.id1.setSelectedTabIndex is not a function
Javascript Error: google is not defined
Javascript Error: google is not defined
Javascript Error: google is not defined
Javascript Error: window.id1.setSelectedTabIndex is not a function
Javascript Error: colab_lib is not defined
Javascript Error: window.id2.setSelectedTabIndex is not a function
Javascript Error: google is not defined
Javascript Error: google is not defined
Javascript Error: window.id2.setSelectedTabIndex is not a function
The following query is stored at Employee_sql variable.
SELECT * FROM (
  
    SELECT
      "Alice" AS name,
      "Product Manager" AS role
   UNION ALL
  
    SELECT
      "Bob" AS name,
      "Engineer" AS role
   UNION ALL
  
    SELECT
      "Caroline" AS name,
      "Engineer" AS role
   UNION ALL
  
    SELECT
      "David" AS name,
      "Data Scientist" AS role
   UNION ALL
  
    SELECT
      "Eve" AS name,
      "Data Scientist" AS role
  
) AS UNUSED_TABLE_NAME  ;
Javascript Error: google is not defined
Javascript Error: google is not defined
Output of this cell has been trimmed on the initial display.
Displaying the first 50 top outputs.
Click on this message to get the complete output.
```

**Reply by EvgSkv** [reply]
Hi @yilinxia 
it looks like Google Cloud didn't install properly for some reason. Not sure what went wrong...
Is it an option for you to use Google Colab? No installation would be needed there.

**Reply by yilinxia** [reply]
@EvgSkv Yeah, I can use Colab. Thank you for the explanation!

**Reply by EvgSkv** [reply]
Good to hear that!
I wish google's toolkit was easier to install...
